Source/Matter/Weyl4WithMatter.impl.hpp

```diff
@@ -0,0 +1,128 @@
+
+    // Add in matter terms to E and B fields
+    add_matter_EB(ebfields, vars, d1, epsilon3_LUU, h_UU, chris);
+
+    // work out the Newman Penrose scalar
+    NPScalar_t<data_t> out =
+        compute_Weyl4(ebfields, vars, d1, d2, h_UU, coords);
+
+    // Write the rhs into the output FArrayBox
+    derive_arrays(i, j, k, m_dcomp)     = out.Real;
+    derive_arrays(i, j, k, m_dcomp + 1) = out.Im;
+}
+
+template <class matter_t>
+template <class data_t>
+AMREX_GPU_DEVICE AMREX_FORCE_INLINE void
+Weyl4WithMatter<matter_t>::add_matter_EB(EBFields_t<data_t> &ebfields,
+                                         const Vars<data_t> &vars,
+                                         const Vars<Tensor<1, data_t>> &d1,
+                                         const Tensor<3, data_t> &epsilon3_LUU,
+                                         const Tensor<2, data_t> &h_UU,
+                                         const chris_t<data_t> &chris) const
+{
+    // Calculate decomposed energy momentum tensor components
+    const auto emtensor = m_matter.compute_emtensor(vars, d1, h_UU, chris.ULL);
+
+    Tensor<2, data_t> Sij_TF = emtensor.Sij;
+    TensorAlgebra::make_trace_free(Sij_TF, vars.h, h_UU);
+
+    // as we made the vacuum expression of Bij explictly symmetric and Eij
+    // explictly trace-free, only Eij has matter terms
+    FOR (i, j)
+    {
+        ebfields.E[i][j] += -4.0 * M_PI * m_G_Newton * Sij_TF[i][j];
+    }
+}
+
+template <class matter_t>
+void Weyl4WithMatter<matter_t>::set_up(int a_state_index)
+{
+
+    int num_ghosts = 2;
+
+    auto &derive_lst     = amrex::AmrLevel::get_derive_lst();
+    const auto &desc_lst = amrex::AmrLevel::get_desc_lst();
+
+    // Add MatterWeyl4 to the derive list
+    derive_lst.add(
+        Weyl4::name, amrex::IndexType::TheCellType(),
+        static_cast<int>(Weyl4::var_names.size()), Weyl4::var_names,
+        Weyl4WithMatter::compute_mf, [=](const amrex::Box &box)
+        { return amrex::grow(box, num_ghosts); }, &amrex::cell_quartic_interp);
+
+    derive_lst.addComponent(Weyl4::name, desc_lst, a_state_index, 0, NUM_VARS);
+}
+
+template <class matter_t>
+void Weyl4WithMatter<matter_t>::compute_mf(amrex::MultiFab &out_mf, int dcomp,
+                                           int ncomp,
+                                           const amrex::MultiFab &src_mf,
+                                           const amrex::Geometry &geomdata,
+                                           amrex::Real /*time*/,
+                                           const int * /*bcrec*/, int /*level*/)
+{
+    const auto &out_arrays = out_mf.arrays();
+    const auto &src_arrays = src_mf.const_arrays();
+
+    GRParmParse pp;
+    std::array<double, AMREX_SPACEDIM> center{};
+    int formulation      = 0;
+    amrex::Real G_Newton = 0;
+
+    pp.get("extraction_center", center);
+    pp.get("formulation", formulation);
+    pp.load("G_newton", G_Newton, 0.0);
```

Comment:
I think we shouldn't use `load` since in theory it can print out some output if
it uses the default argument (e.g. if the example doesn't add the default to the
`ParmParse` table).
```suggestion
    pp.queryAdd("G_newton", G_Newton);
```